Scheduler: Add query fairness control across multiple actors within a tenant

chaudum opened the pull request:
**What this PR does / why we need it**:

This PR wires up the scheduler with the hierarchical queues.
It is the last PR to implement https://github.com/grafana/loki/pull/8585.

When these changes are in place, the client performing query requests can control their QoS (query fairness) using the `X-Actor-Path` HTTP header. This header controls in which sub-queue of the tenant's scheduler queue the query request is enqueued. The place within the hierarchy where it is enqueued defines the probability with which the request gets dequeued.

A common use-case for this QoS control is giving each Grafana user within a tenant their fair share of query execution time.

Any documentation is still missing and will be provided by follow-up PRs.

**Special notes for your reviewer**:

```console
$ gotest -count=1 -v ./pkg/scheduler/queue/... -test.run=TestQueryFairness
=== RUN   TestQueryFairness
=== RUN   TestQueryFairness/use_hierarchical_queues_=_false
    dequeue_qos_test.go:109: duration actor a 2.007765568s
    dequeue_qos_test.go:109: duration actor b 2.209088331s
    dequeue_qos_test.go:112: total duration 2.209280772s
=== RUN   TestQueryFairness/use_hierarchical_queues_=_true
    dequeue_qos_test.go:109: duration actor b 605.283144ms
    dequeue_qos_test.go:109: duration actor a 2.270931324s
    dequeue_qos_test.go:112: total duration 2.271108551s
--- PASS: TestQueryFairness (4.48s)
    --- PASS: TestQueryFairness/use_hierarchical_queues_=_false (2.21s)
    --- PASS: TestQueryFairness/use_hierarchical_queues_=_true (2.27s)
PASS
ok  	github.com/grafana/loki/pkg/scheduler/queue	4.491s
```

```console
$ gotest -count=5 -v ./pkg/scheduler/queue/... -bench=Benchmark -test.run=^$ -benchtime=10000x -benchmem
goos: linux
goarch: amd64
pkg: github.com/grafana/loki/pkg/scheduler/queue
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
BenchmarkGetNextRequest
BenchmarkGetNextRequest/without_sub-queues
BenchmarkGetNextRequest/without_sub-queues-8         	   10000	     29337 ns/op	    1600 B/op	     100 allocs/op
BenchmarkGetNextRequest/without_sub-queues-8         	   10000	     21348 ns/op	    1600 B/op	     100 allocs/op
BenchmarkGetNextRequest/without_sub-queues-8         	   10000	     21595 ns/op	    1600 B/op	     100 allocs/op
BenchmarkGetNextRequest/without_sub-queues-8         	   10000	     21189 ns/op	    1600 B/op	     100 allocs/op
BenchmarkGetNextRequest/without_sub-queues-8         	   10000	     21602 ns/op	    1600 B/op	     100 allocs/op
BenchmarkGetNextRequest/with_1_level_of_sub-queues
BenchmarkGetNextRequest/with_1_level_of_sub-queues-8 	   10000	     33770 ns/op	    2400 B/op	     200 allocs/op
BenchmarkGetNextRequest/with_1_level_of_sub-queues-8 	   10000	     33596 ns/op	    2400 B/op	     200 allocs/op
BenchmarkGetNextRequest/with_1_level_of_sub-queues-8 	   10000	     34432 ns/op	    2400 B/op	     200 allocs/op
BenchmarkGetNextRequest/with_1_level_of_sub-queues-8 	   10000	     33760 ns/op	    2400 B/op	     200 allocs/op
BenchmarkGetNextRequest/with_1_level_of_sub-queues-8 	   10000	     33664 ns/op	    2400 B/op	     200 allocs/op
BenchmarkGetNextRequest/with_2_levels_of_sub-queues
BenchmarkGetNextRequest/with_2_levels_of_sub-queues-8         	   10000	     71405 ns/op	    3200 B/op	     300 allocs/op
BenchmarkGetNextRequest/with_2_levels_of_sub-queues-8         	   10000	     59472 ns/op	    3200 B/op	     300 allocs/op
BenchmarkGetNextRequest/with_2_levels_of_sub-queues-8         	   10000	    117163 ns/op	    3200 B/op	     300 allocs/op
BenchmarkGetNextRequest/with_2_levels_of_sub-queues-8         	   10000	    106505 ns/op	    3200 B/op	     300 allocs/op
BenchmarkGetNextRequest/with_2_levels_of_sub-queues-8         	   10000	     64374 ns/op	    3200 B/op	     300 allocs/op
BenchmarkQueueRequest
BenchmarkQueueRequest-8                                       	   10000	    168391 ns/op	  320588 B/op	    1156 allocs/op
BenchmarkQueueRequest-8                                       	   10000	    166203 ns/op	  320587 B/op	    1156 allocs/op
BenchmarkQueueRequest-8                                       	   10000	    149518 ns/op	  320584 B/op	    1156 allocs/op
BenchmarkQueueRequest-8                                       	   10000	    219776 ns/op	  320583 B/op	    1156 allocs/op
BenchmarkQueueRequest-8                                       	   10000	    185198 ns/op	  320597 B/op	    1156 allocs/op
PASS
ok  	github.com/grafana/loki/pkg/scheduler/queue	64.648s
```

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

chaudum commented on pkg/lokifrontend/frontend/v1/frontend.go:
```diff
@@ -327,7 +327,7 @@ func (f *Frontend) queueRequest(ctx context.Context, req *request) error {
 	f.activeUsers.UpdateUserTimestamp(joinedTenantID, now)
 
-	err = f.requestQueue.Enqueue(joinedTenantID, req, maxQueriers, nil)
+	err = f.requestQueue.Enqueue(joinedTenantID, nil, req, maxQueriers, nil)
```
Do we want to enable hierarchical queuing in the v1 frontend (single binary!) as well? If so, we probably need to change the `max_queue_hierarchy_levels` setting from being a scheduler setting to being a frontend setting (which is not great).

jeschkies replied:
It's say no.

chaudum commented on pkg/scheduler/scheduler.go:
```diff
@@ -61,6 +62,12 @@ const (
 	ringCheckPeriod = 3 * time.Second
+
+	// ActorPathHeader is the name of the header used to enqueue requests in hierarchical queues
+	ActorPathHeader = "x-actor-path"
```
I wonder, whether we should start using `X-Loki-*` headers consistently, like here for the query limits: https://github.com/grafana/loki/pull/8727/files#diff-a19fa86227cdb8ba8ac4d877d8f8fa4337c00ab92594fae43cdae5b9b5c10a96R12

```suggestion
	ActorPathHeader = "x-loki-actor-path"
```

chaudum commented on pkg/querier/queryrange/codec.go:
```diff
@@ -270,19 +271,34 @@ func (Codec) DecodeRequest(_ context.Context, r *http.Request, forwardHeaders []
 		header.Set(string(httpreq.QueryTagsHTTPHeader), queryTags)
 	}
 
+	actor := httpreq.ExtractHeader(ctx, httpreq.LokiActorPathHeader)
+	if actor != "" {
+		header.Set(httpreq.LokiActorPathHeader, actor)
+	}
```
This is not strictly necessary.

chaudum replied:
This sets the x-loki-actor-path HTTP header again when converting a LokiRequest back to a HTTP requests. That would ensure the header is also present in the httpgrcp.Request that is sent from the Frontend to the Scheduler. However, by adding the additional `QueuePath` field to the gRPC message, it's never read out in the scheduler.

jeschkies commented on pkg/lokifrontend/frontend/v2/frontend.go:
```diff
@@ -188,7 +190,7 @@ func (f *Frontend) RoundTripGRPC(ctx context.Context, req *httpgrpc.HTTPRequest)
 		return nil, err
 	}
-	userID := tenant.JoinTenantIDs(tenantIDs)
+	tenantID := tenant.JoinTenantIDs(tenantIDs)
```
👍

jeschkies commented on pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go:
```diff
@@ -286,7 +286,8 @@ func (w *frontendSchedulerWorker) schedulerLoop(loop schedulerpb.SchedulerForFro
 				QueryID:         req.queryID,
-				UserID:          req.userID,
+				UserID:          req.tenantID,
+				QueuePath:       req.actor,
```
Hm. Should we stick to the same name?

jeschkies commented on pkg/scheduler/scheduler.go:
```diff
@@ -408,14 +408,24 @@ func (s *Scheduler) enqueueRequest(frontendContext context.Context, frontendAddr
+	if s.cfg.MaxQueueHierarchyLevels > 0 {
+		queuePath = msg.QueuePath
+		// trim to the max allowed levels
+		// TODO(chaudum): Would it be better to return an error?
```
I'd say so.

chaudum commented on pkg/util/httpreq/headers.go:
```diff
@@ -0,0 +1,44 @@
+	return middleware.Func(func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			for _, h := range headers {
+				value := req.Header.Get(LokiActorPathHeader)
```
That line should be `value := req.Header.Get(h)`

chaudum commented on pkg/querier/queryrange/limits.go:
```diff
@@ -363,10 +362,6 @@ func (rt limitedRoundTripper) do(ctx context.Context, r queryrangebase.Request)
 		return nil, err
 	}
 
-	if err := user.InjectOrgIDIntoHTTPRequest(ctx, request); err != nil {
-		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
-	}
-
```
Ok, will revert this change.

owen-d commented on pkg/scheduler/queue/leafqueue.go:
```diff
@@ -78,7 +75,7 @@ func (q *LeafQueue) Dequeue() Request {
 
 	// only if there are no items queued in the local queue, dequeue from sub-queues
-	maxIter := q.mapping.Len()
+	maxIter := len(q.mapping.keys)
```
Wasn't the old `q.mapping.Len()` correct since it subtracted empty keys?

chaudum replied:
Imagine `keys=[x, x, "foo", "bar", x]` where x marks the empty slot. When you are at index=3 and want to get the next items.
If you use `Len()`, then you would only iterate to index=0, which is empty and therefore would not return a sub queue.

owen-d commented on pkg/scheduler/queue/leafqueue.go:
```diff
@@ -88,6 +85,8 @@ func (q *LeafQueue) Dequeue() Request {
 			}
 			q.mapping.Remove(subq.name)
+		} else {
+			q.current++
```
Why do we need to update `q.current` if there wasn't a found sub queue?

owen-d reviewed (commented):
A couple questions, but looking pretty good

chaudum commented on pkg/scheduler/schedulerpb/scheduler.proto:
```diff
@@ -77,6 +77,8 @@ message FrontendToScheduler {
   httpgrpc.HTTPRequest httpRequest = 5;
   bool statsEnabled = 6;
+  // Path to queue to which the request will be enqueued.
+  repeated string queuePath = 7;
```
@jeschkies raised a good point: The scheduler should be agnostic of the `Request` object that passes the scheduler.
By encoding the queuePath into the header of httpgrpc.Request, the scheduler knows the implementation of the `Request`.

chaudum commented on pkg/scheduler/queue/leafqueue.go:
```diff
@@ -78,7 +75,7 @@ func (q *LeafQueue) Dequeue() Request {
 
 	// only if there are no items queued in the local queue, dequeue from sub-queues
-	maxIter := q.mapping.Len()
+	maxIter := len(q.mapping.keys)
```
Yes, that makes a lot of sense. I addressed it in https://github.com/grafana/loki/pull/8752/commits/f32c3f0c8345f83b690483c2c2a1b64f9139db1d

JStickler reviewed (approved):
[docs team] LGTM.

"Any documentation is still missing and will be provided by follow-up PRs."  Tag me when you write it.

chaudum commented:
Rebased with `main` and added https://github.com/grafana/loki/pull/8752/commits/c5dbad7c80be79b273ef87d282b3910a4fe08525 (which was previously reverted on `main`).

owen-d commented:
Hm, I'm worried about the negative queue length. Let's figure out why

chaudum commented on pkg/scheduler/queue/leafqueue.go:
```diff
@@ -0,0 +1,155 @@
+		}
+
+		subq, err := q.mapping.GetNext(q.current)
+		if err == ErrOutOfBounds {
```
It means that we reached the end of the sub-queue list and need to start with start index again.

chaudum commented on pkg/scheduler/queue/dequeue_qos_test.go:
```diff
@@ -0,0 +1,183 @@
+			wg.Wait()
+
+			require.Equal(t, total, responseCount.Load())
+			durations.Range(func(k, v any) bool {
+				t.Log("duration actor", k, v)
+				return true
+			})
+			t.Log("total duration", time.Since(start))
```
It's rather a benchmark than a test, really. Maybe I should change it to a Benchmark.

chaudum commented on pkg/scheduler/queue/leafqueue.go:
```diff
@@ -0,0 +1,152 @@
+// LeafQueue is an hierarchical queue implementation where each sub-queue
+// has the same guarantees to be chosen from.
+// Each queue has also a local queue, which gets chosen with equal preference as the sub-queues.
+type LeafQueue struct {
```
Either once the PR is approved and rebased, or later in a second PR

owen-d reviewed (approved):
skimmed the diffs since last review, lgtm
